frontend/expr_inferrer.py

```diff
@@ -0,0 +1,372 @@
+	 - Await(expr value) --> Python 3.5+
+	 - Yield(expr? value)
+     - Compare(expr left, cmpop* ops, expr* comparators)
+     - Name(identifier id, expr_context ctx)
+     - FormattedValue(expr value, int? conversion, expr? format_spec) --> Python 3.6+
+     - JoinedStr(expr* values) --> Python 3.6+
+     - ListComp(expr elt, comprehension* generators)
+     - SetComp(expr elt, comprehension* generators)
+     - DictComp(expr key, expr value, comprehension* generators)
+
+     TODO:
+     - Lambda(arguments args, expr body)
+     - GeneratorExp(expr elt, comprehension* generators)
+     - YieldFrom(expr value)
+     - Call(expr func, expr* args, keyword* keywords)
+     - Attribute(expr value, identifier attr, expr_co0ontext ctx)
+     - Starred(expr value, expr_context ctx)
+
+     TODO:
+     - Infer (or narrow) types based on context. For example, in the following expression:
+        (x + 2) * 3
+        if the type of x is still not inferred (or inferred to be a union type), we can infer (narrow) that x
+        is a numeric value (TBool, TInt or TFloat)
+        Affected inference functions:
+            - infer_binary_operation
+            - infer_unary_operation
+            - infer_subscript
+            - infer_compare
+"""
+
+import ast, sys
+from i_types import *
+from context import Context
+from abc import ABCMeta, abstractmethod
+from exceptions import HomogeneousTypesConflict
+
+def infer_numeric(node):
+    """Infer the type of a numeric node"""
+    if type(node.n) == int:
+        return TInt()
+    if type(node.n) == float:
+        return TFloat()
+
+def infer_list(node, context):
+    """Infer the type of a homogeneous list
+
+    Returns: TList(Type t), where t is the type of the list elements
+    """
+    if len(node.elts) == 0:
+        return TList(TNone())
+    list_type = infer(node.elts[0], context)
+    for i in range(1, len(node.elts)):
+        cur_type = infer(node.elts[i], context)
+        if list_type.is_subtype(cur_type): # get the most generic type
+            list_type = cur_type
+        elif not cur_type.is_subtype(list_type): # make sure the list is homogeneous
+            raise HomogeneousTypesConflict(list_type.get_name(), cur_type.get_name())
+    return TList(list_type)
+
+def infer_dict(node, context):
+    """Infer the type of a dictionary with homogeneous key set and value set
+
+    Returns: TDictionary(Type k_t, Type v_t), where:
+            k_t is the type of dictionary keys
+            v_t is the type of dictionary values
+    """
+    if len(node.keys) == 0:
+        return TDictionary(TNone(), TNone())
+    keys = node.keys
+    values = node.values
+    keys_type = infer(keys[0], context)
+    values_type = infer(values[0], context)
+
+    for i in range(1, len(keys)):
+        cur_key_type = infer(keys[i], context)
+        cur_value_type = infer(values[i], context)
+
+        if keys_type.is_subtype(cur_key_type): # get the most generic keys type
```

Comment:
This is essentially the same code as in infer_list and infer_set, you could probably extract that into its own function.